x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/user/TransportHasPrivilegesAction.java

```diff
@@ -139,4 +196,17 @@ private static boolean testIndex(Automaton checkIndex, Automaton roleIndex) {
     private static boolean testPrivilege(Privilege checkPrivilege, Automaton roleAutomaton) {
         return Operations.subsetOf(checkPrivilege.getAutomaton(), roleAutomaton);
     }
+
+    private boolean testResourceMatch(String application, String checkResource, String checkPrivilegeName, Role userRole,
+                                      Collection<ApplicationPrivilege> privileges) {
+        final Set<String> nameSet = Collections.singleton(checkPrivilegeName);
+        final ApplicationPrivilege checkPrivilege = ApplicationPrivilege.get(application, nameSet, privileges);
+        assert checkPrivilege.getApplication().equals(application)
+            : "Privilege " + checkPrivilege + " should have application " + application;
+        assert checkPrivilege.name().equals(nameSet)
+            : "Privilege " + checkPrivilege + " should have name " + nameSet;
+
+        return userRole.application().grants(checkPrivilege, checkResource);
```

Comment:
Pertaining to the 'Role cache clean saga':
I understand the `checkPrivilege` is freshly baked (not cached), but the `userRole` application permissions are built from cached application privileges. Since we already pull the latest privileges here, do you think it would be possible to refactor `Role` so that it only holds the names of permissions.
I haven't fully grasped all the machinery so this might be nonsense...

Reply:
I think it might be possible.
At the moment we've definitely got the wrong tradeoffs. We're loading privileges from the index every time, and compiling the automata for each of them, but then using the cahced copy on the role.
I'll come back to this after the refactor, because that will change things a bit (the stuff we load from the index won't be an actual privilege object)